### ISSUE TITLE
3.21.0-0.0.pre encounters asserts on Linux when showing GTK widgets only after plugin registration

### ISSUE THREAD

**Issue by Craftplacer** [flutter/flutter#144873]
### Steps to reproduce

1. Create new Flutter project with a plugin that interfaces with native code
2. Move code for showing GTK widgets after `fl_register_plugins`
3. Try to interface with plugin
4. Observe the application failing to do so

### Expected results

Ability to use the plugins

### Actual results

Flutter complains about missing platform channels and throws assertion warnings in the standard output

### Code sample

From https://pub.dev/packages/handy_window

```diff 
diff --git a/linux/my_application.cc b/linux/my_application.cc
index fa74baa..3133755 100644
--- a/linux/my_application.cc
+++ b/linux/my_application.cc
@@ -48,17 +48,17 @@ static void my_application_activate(GApplication* application) {
   }
 
   gtk_window_set_default_size(window, 1280, 720);
-  gtk_widget_show(GTK_WIDGET(window));
 
   g_autoptr(FlDartProject) project = fl_dart_project_new();
   fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);
 
   FlView* view = fl_view_new(project);
-  gtk_widget_show(GTK_WIDGET(view));
   gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));
 
   fl_register_plugins(FL_PLUGIN_REGISTRY(view));
 
+  gtk_widget_show(GTK_WIDGET(window));
+  gtk_widget_show(GTK_WIDGET(view));
   gtk_widget_grab_focus(GTK_WIDGET(view));
 }
```

```dart
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

void main() => runApp(const MainApp());

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: FilledButton(
            onPressed: () => launchUrl(Uri.https('flutter.dev')),
            child: const Text('Open site'),
          ),
        ),
      ),
    );
  }
}

```

### Logs

<details open><summary>Logs</summary>

```console

** (app:3205): CRITICAL **: 11:04:29.166: FlBinaryMessenger *fl_engine_get_binary_messenger(FlEngine *): assertion 'FL_IS_ENGINE(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlTextureRegistrar *fl_engine_get_texture_registrar(FlEngine *): assertion 'FL_IS_ENGINE(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlPluginRegistrar *fl_plugin_registrar_new(FlView *, FlBinaryMessenger *, FlTextureRegistrar *): assertion 'FL_IS_BINARY_MESSENGER(messenger)' failed

(app:3205): GLib-GObject-CRITICAL **: 11:04:29.166: g_object_ref: assertion 'G_IS_OBJECT (object)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlBinaryMessenger *fl_plugin_registrar_get_messenger(FlPluginRegistrar *): assertion 'FL_IS_PLUGIN_REGISTRAR(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlMethodChannel *fl_method_channel_new(FlBinaryMessenger *, const gchar *, FlMethodCodec *): assertion 'FL_IS_BINARY_MESSENGER(messenger)' failed

** (app:3205): CRITICAL **: 11:04:29.166: void fl_method_channel_set_method_call_handler(FlMethodChannel *, FlMethodChannelMethodCallHandler, gpointer, GDestroyNotify): assertion 'FL_IS_METHOD_CHANNEL(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlBinaryMessenger *fl_engine_get_binary_messenger(FlEngine *): assertion 'FL_IS_ENGINE(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlTextureRegistrar *fl_engine_get_texture_registrar(FlEngine *): assertion 'FL_IS_ENGINE(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlPluginRegistrar *fl_plugin_registrar_new(FlView *, FlBinaryMessenger *, FlTextureRegistrar *): assertion 'FL_IS_BINARY_MESSENGER(messenger)' failed

(app:3205): GLib-GObject-CRITICAL **: 11:04:29.166: g_object_ref: assertion 'G_IS_OBJECT (object)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlBinaryMessenger *fl_plugin_registrar_get_messenger(FlPluginRegistrar *): assertion 'FL_IS_PLUGIN_REGISTRAR(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlMethodChannel *fl_method_channel_new(FlBinaryMessenger *, const gchar *, FlMethodCodec *): assertion 'FL_IS_BINARY_MESSENGER(messenger)' failed

** (app:3205): CRITICAL **: 11:04:29.166: void fl_method_channel_set_method_call_handler(FlMethodChannel *, FlMethodChannelMethodCallHandler, gpointer, GDestroyNotify): assertion 'FL_IS_METHOD_CHANNEL(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlBinaryMessenger *fl_engine_get_binary_messenger(FlEngine *): assertion 'FL_IS_ENGINE(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlTextureRegistrar *fl_engine_get_texture_registrar(FlEngine *): assertion 'FL_IS_ENGINE(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlPluginRegistrar *fl_plugin_registrar_new(FlView *, FlBinaryMessenger *, FlTextureRegistrar *): assertion 'FL_IS_BINARY_MESSENGER(messenger)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlView *fl_plugin_registrar_get_view(FlPluginRegistrar *): assertion 'FL_IS_PLUGIN_REGISTRAR(self)' failed

(app:3205): Gtk-CRITICAL **: 11:04:29.166: gtk_widget_get_toplevel: assertion 'GTK_IS_WIDGET (widget)' failed

(app:3205): Gtk-CRITICAL **: 11:04:29.166: gtk_window_get_screen: assertion 'GTK_IS_WINDOW (window)' failed

(app:3205): Gdk-CRITICAL **: 11:04:29.166: gdk_screen_is_composited: assertion 'GDK_IS_SCREEN (screen)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlBinaryMessenger *fl_engine_get_binary_messenger(FlEngine *): assertion 'FL_IS_ENGINE(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlTextureRegistrar *fl_engine_get_texture_registrar(FlEngine *): assertion 'FL_IS_ENGINE(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlPluginRegistrar *fl_plugin_registrar_new(FlView *, FlBinaryMessenger *, FlTextureRegistrar *): assertion 'FL_IS_BINARY_MESSENGER(messenger)' failed

(app:3205): GLib-GObject-CRITICAL **: 11:04:29.166: g_object_ref: assertion 'G_IS_OBJECT (object)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlBinaryMessenger *fl_plugin_registrar_get_messenger(FlPluginRegistrar *): assertion 'FL_IS_PLUGIN_REGISTRAR(self)' failed

** (app:3205): CRITICAL **: 11:04:29.166: FlMethodChannel *fl_method_channel_new(FlBinaryMessenger *, const gchar *, FlMethodCodec *): assertion 'FL_IS_BINARY_MESSENGER(messenger)' failed

** (app:3205): CRITICAL **: 11:04:29.166: void fl_method_channel_set_method_call_handler(FlMethodChannel *, FlMethodChannelMethodCallHandler, gpointer, GDestroyNotify): assertion 'FL_IS_METHOD_CHANNEL(self)' failed

```

</details>

<details open><summary>Exception</summary>

```console

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: MissingPluginException(No implementation found for method launch on channel plugins.flutter.io/url_launcher_linux)
#0      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:332:7)
<asynchronous suspension>
#1      UrlLauncherLinux.launchUrl (package:url_launcher_linux/url_launcher_linux.dart:47:13)
<asynchronous suspension>

```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel master, 3.21.0-2.0.pre.63, on Artix Linux 6.7.4-artix1-1, locale en_US.UTF-8)
    • Flutter version 3.21.0-2.0.pre.63 on channel master at /home/nyanya/.local/lib/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f733e3a441 (3 hours ago), 2024-03-09 15:50:23 +0800
    • Engine revision 196132ffe2
    • Dart version 3.4.0 (build 3.4.0-219.0.dev)
    • DevTools version 2.33.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /home/nyanya/.local/lib/android-sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /opt/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /usr/bin/chromium

[✓] Linux toolchain - develop for Linux desktop
    • clang version 16.0.6
    • cmake version 3.28.3
    • ninja version 1.11.1
    • pkg-config version 2.1.0

[!] Android Studio (version unknown)
    • Android Studio at /opt/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.3)
    • IntelliJ at /home/nyanya/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate
    • Flutter plugin version 78.1.1
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Artix Linux 6.7.4-artix1-1
    • Chrome (web)    • chrome • web-javascript • Chromium 122.0.6261.39 Arch Linux

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.

```

</details>

**Comment by stuartmorgan** [comment]
Closing as a duplicate of https://github.com/flutter/flutter/issues/118269

**Comment by stuartmorgan** [comment]
Re-opening since I may have misunderstood the issue here; looking again, your description ("Flutter complains about missing platform channels") and your log output don't match.

Could you please update your Steps to Reproduce section to be clearer and more detailed? In particular:
* > Create new Flutter project with a plugin that interfaces with native code
 
  This is not enough information for someone to reproduce it. What one plugin are you adding to a new, default project to reproduce this?
* > Try to interface with the plugin
 
  Interface how?
* It's not clear to me if your Code Sample section is part of the reproductions steps, or a suggested fix. I read this as the latter initially; if changing the native project is part of the repro steps, the repro steps should be clear about that.

**Comment by huycozy** [comment]
I summarize the `steps to reproduce` as below:

1. Create a new Flutter project. Then add a plugin, eg: `url_launcher`
2. Update `main.dart` code as above. 
3. Update `/linux/my_application.cc` as above (move these 2 lines of `gtk_widget_show` after `fl_register_plugins`)

Before (default code when created project): 

```c
  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
```

After:

```c
  gtk_window_set_default_size(window, 1280, 720);

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));
  gtk_widget_show(GTK_WIDGET(view));
  gtk_widget_show(GTK_WIDGET(window));

  gtk_widget_grab_focus(GTK_WIDGET(view));
```
4. Run Linux app target

The issue only occurs on master channel but not on stable `3.19.3`. 

<details>
<summary>flutter doctor -v (stable & master)</summary>

```bash
[✓] Flutter (Channel stable, 3.19.3, on Ubuntu 22.04.3 LTS 6.5.0-21-generic, locale en_US.UTF-8)
    • Flutter version 3.19.3 on channel stable at /home/huynq/Documents/Working/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ba39319843 (4 days ago), 2024-03-07 15:22:21 -0600
    • Engine revision 2e4ba9c6fb
    • Dart version 3.3.1
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc3)
    • Android SDK at /home/huynq/Android/Sdk/
    • Platform android-33, build-tools 34.0.0-rc3
    • Java binary at: /snap/android-studio/current/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1.1
    • cmake version 3.22.1
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 2021.1)
    • Android Studio at /home/huynq/Documents/android-studio
    • Flutter plugin version 67.0.1
    • Dart plugin version 211.7817
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[!] Android Studio (version unknown)
    • Android Studio at /snap/android-studio/current
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • android-studio-dir = /snap/android-studio/current
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.84.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.84.0

[✓] VS Code (version 1.87.0)
    • VS Code at /snap/code/current/usr/share/code
    • Flutter extension version 3.84.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 22.04.3 LTS 6.5.0-21-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 120.0.6099.199

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

```bash
[!] Flutter (Channel master, 3.21.0-2.0.pre.73, on Ubuntu 22.04.3 LTS 6.5.0-21-generic, locale en_US.UTF-8)
    • Flutter version 3.21.0-2.0.pre.73 on channel master at /home/huynq/Documents/Working/flutter_master
    ! Warning: `flutter` on your path resolves to /home/huynq/Documents/Working/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /home/huynq/Documents/Working/flutter_master. Consider adding /home/huynq/Documents/Working/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /home/huynq/Documents/Working/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /home/huynq/Documents/Working/flutter_master. Consider adding /home/huynq/Documents/Working/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 8e12e133aa (32 minutes ago), 2024-03-11 04:57:34 -0400
    • Engine revision 210f84eb81
    • Dart version 3.4.0 (build 3.4.0-220.0.dev)
    • DevTools version 2.33.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc3)
    • Android SDK at /home/huynq/Android/Sdk/
    • Platform android-33, build-tools 34.0.0-rc3
    • Java binary at: /snap/android-studio/current/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1.1
    • cmake version 3.22.1
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 2021.1)
    • Android Studio at /home/huynq/Documents/android-studio
    • Flutter plugin version 67.0.1
    • Dart plugin version 211.7817
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[!] Android Studio (version unknown)
    • Android Studio at /snap/android-studio/current
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • android-studio-dir = /snap/android-studio/current
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.84.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.84.0

[✓] VS Code (version 1.87.0)
    • VS Code at /snap/code/current/usr/share/code
    • Flutter extension version 3.84.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 22.04.3 LTS 6.5.0-21-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 120.0.6099.199

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```
</details>

**Comment by gspencergoog** [comment]
cc @robert-ancell

**Comment by gspencergoog** [comment]
Fixing this may also be needed for multi-view implementation.

**Comment by zoechi** [comment]
I just run into this with https://github.com/Yubico/yubioath-flutter on NixOS

**Comment by ZG-123** [comment]
Hello, may I ask if this problem has been solved? I also encountered the same problem, after packaging on Linux, I can not open, and the error is also this.

**Comment by spydon** [comment]
At Canonical we are heavily affected by this bug, I think there is a `customer: canonical` tag, or have I made that up?  :thinking:
